README.md
# f1-telemetry-cli
![CircleCI](https://img.shields.io/circleci/build/github/chaseconey/f1-telemetry-cli)
![npm (tag)](https://img.shields.io/npm/v/f1-telemetry-cli/latest)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)


A simple CLI for tracking and processing F1 game UDP data (currently supports F1 2020).

Companion CLI for the [F1 Racing Portal](https://github.com/chaseconey/frl).

## Overview

This CLI uses the in-game UDP packet stream to parse and process the race data as it comes in. This data is then persisted to a flat json file on interval for later retrieval/processing. There are many other solutions out there that accomplish this in different ways. The approach of this package was to try and be very selective about what data is stored to try and keep the process as lightweight as possible. If you are looking to collect as much data as possible for more powerful post-processing capabilities, you might want to check out [this project](https://pypi.org/project/f1-2020-telemetry/).

### Tracked Data

Here is a high level idea of the kinds of data that this CLI records:

- Session ID
- Driver data such as racing number and team
- Snapshot lap data for tracking fastest lap, fastest sectors, and driver status
- End of race classification data for all drivers such as fastest lap time, points, pit stops, penalties (basically everything you see on the final screen after a race/quali)
- Custom data such as best lap tire (this is not provided in lap or classification data)

For a full example of a completed race, check out [the stub here](stubs/final-race.json).

## Installation

```
npm install -g f1-telemetry-cli
```

## Usage

```
f1-telemetry help
f1-telemetry [command]

Commands:
  f1-telemetry record          Start recording F1 2020 sessions
  f1-telemetry process <file>  Process session data for specific session
  f1-telemetry session         Get live session data

Options:
  --version  Show version number                                       [boolean]
  --help     Show help                                                 [boolean]
```

### Recorder

```
f1-telemetry record

Start recording F1 2020 sessions

Options:
      --version          Show version number                           [boolean]
      --help             Show help                                     [boolean]
  -p, --path             Path to write file to                    [default: "."]
      --write-frequency  Time in ms to wait between flushes to file
                                                                 [default: 2500]
      --prefix           Name to prefix written file with          [default: ""]
      --port             UDP port to listen on                  [default: 20777]
  -f, --forward-address  IP Address to forward UDP packet stream to
```

This command will record the active UDP data stream from the game and save the important bits and bobs in a very lightweight json file. This can be extremely useful for collecting data for things like league races.

### Processor

```
f1-telemetry process <file>

Process session data for specific session

Options:
      --version  Show version number                                   [boolean]
      --help     Show help                                             [boolean]
  -p, --path     Path to write file to                            [default: "."]
  -n, --name     Name of the processed file               [default: "processed"]
```

This command will further process the raw json data and put it into a slightly more usable format.

This command also augments the original data with some calculated data that could be useful such as:

- **Driver Fastest Lap Delta** - the difference between each driver's fastest lap and the race's fastest lap
- **Driver Fastest Lap Sector Deltas** - the difference between each driver's fastest sector and the session's fastest lap's fastest sector (not the overall fastest sector)

It will then organize this data sorted by the finishing race position and keyed by the driver's racing number.

For a full example of a completed race, check out [the stub here](stubs/final-race.json).

### Session

![](/img/session.png)

```
f1-telemetry session

Get live session data

Options:
      --version       Show version number                              [boolean]
      --help          Show help                                        [boolean]
      --port          UDP port to listen on                     [default: 20777]
      --events-shown  Number of events to show in events table      [default: 5]
  -m, --map-file      Path to csv driver map (racing number, driver name)
                                                                 [default: null]
```

This command can be used to stream session information to your console. This is a great tool for getting at-a-glance information as a broadcaster or spectator.

To correlate real-time events to drivers, you can pass a path to a csv with driver racing numbers to names like so:

```
2,Rawrocopter
21,Redbaron
```

Now everytime an event comes in for racing number `2`, `Rawrocopter` will be shown instead.



CHANGELOG.md
## [1.2.1](https://github.com/chaseconey/f1-telemetry-cli/compare/v1.2.0...v1.2.1) (2021-09-04)


### Bug Fixes

* pin client version to old game ([3bbfe68](https://github.com/chaseconey/f1-telemetry-cli/commit/3bbfe688d8f3b573d34650b4a5b6b1135929925f))

# [1.2.0](https://github.com/chaseconey/f1-telemetry-cli/compare/v1.1.0...v1.2.0) (2021-04-20)


### Features

* add udp forwarding support ([9e22ba9](https://github.com/chaseconey/f1-telemetry-cli/commit/9e22ba9016a0a9dc2778a30510fbd14c4913d044))

# [1.1.0](https://github.com/chaseconey/f1-telemetry-cli/compare/v1.0.2...v1.1.0) (2021-04-05)


### Features

* add session command ([ad9e8a8](https://github.com/chaseconey/f1-telemetry-cli/commit/ad9e8a82b3ef011ff3721f27fe8c947fd51e02c6))

## [1.0.2](https://github.com/chaseconey/f1-telemetry-cli/compare/v1.0.1...v1.0.2) (2021-04-04)


### Bug Fixes

* add bin to package.json ([4d73120](https://github.com/chaseconey/f1-telemetry-cli/commit/4d7312084ef1cbcf9f31ccc30f234f3b8c2c8ac4))

## [1.0.1](https://github.com/chaseconey/f1-telemetry-cli/compare/v1.0.0...v1.0.1) (2021-04-04)


### Bug Fixes

* make entry executable ([10b2927](https://github.com/chaseconey/f1-telemetry-cli/commit/10b2927cd302562775c949a3d6df89c6cb8aa919))

# 1.0.0 (2021-04-04)


### Bug Fixes

* semantic release main branch ([73d35f6](https://github.com/chaseconey/f1-telemetry-cli/commit/73d35f6e32f70f00b463254e1bb883339387d568))
* semantic-release plugins and repo url ([5813653](https://github.com/chaseconey/f1-telemetry-cli/commit/5813653301c6366b0e595d35641f8d50d6412163))
* upgrade node test image ([2fe74d4](https://github.com/chaseconey/f1-telemetry-cli/commit/2fe74d43eec3d44b41b2fdb7f113f546086c4e64))


### Features

* add ci ([819815e](https://github.com/chaseconey/f1-telemetry-cli/commit/819815e592f5241c9ee6343d3208d7b6412999fe))
* add semantic-release ([07f5612](https://github.com/chaseconey/f1-telemetry-cli/commit/07f5612df67918017ae5725341efb794eb2f1a2c))


cmds/record.js
const { F1TelemetryClient, constants } = require("f1-telemetry-client");
const { PACKETS } = constants;
const fs = require("fs");
const { trackParticipants } = require("../lib/participants");
const { trackLapData } = require("../lib/laps");
const { trackFinalClassification } = require("../lib/finalClassification");
const { trackEvents } = require("../lib/events");

let raceData = {
  sessionId: null,
  data: {},
};
let lastWrite = null;
let port = 20777;
let writeFrequency = 2500;
let prefix = "";
let path = ".";

function persistData(data, force = false) {
  if (data && (force || Date.now() - lastWrite > writeFrequency)) {
    console.info(`Persisting data to ${prefix}race-${data.sessionId}.json`);
    fs.writeFileSync(
      `${path}/${prefix}race-${data.sessionId}.json`,
      JSON.stringify(data),
      function (err) {
        if (err) return console.log(err);
      }
    );

    lastWrite = Date.now();
  }
}

function resetData() {
  raceData = {
    sessionId: null,
    data: {},
  };
  lastWrite = null;
}

exports.command = "record";

exports.describe = "Start recording F1 2020 sessions";

exports.builder = {
  path: {
    default: ".",
    alias: "p",
    desc: "Path to write file to",
  },
  "write-frequency": {
    default: 2500,
    desc: "Time in ms to wait between flushes to file",
  },
  prefix: {
    default: "",
    desc: "Name to prefix written file with",
  },
  port: {
    default: 20777,
    desc: "UDP port to listen on",
  },
  "forward-address": {
    alias: "f",
    desc: "IP Address to forward UDP packet stream to",
  },
};

exports.handler = function (argv) {
  writeFrequency = argv.writeFrequency;
  port = argv.port;
  prefix = argv.prefix;
  path = argv.path;

  let config = {
    port,
  };

  if (argv.forwardAddress) {
    config.forwardAddresses = [{ port, ip: argv.forwardAddress }];
  }

  let carStatus = {};

  const client = new F1TelemetryClient(config);
  // client.on(PACKETS.motion, console.log);
  // client.on(PACKETS.carSetups, console.log);
  // client.on(PACKETS.session, (s) => console.dir(s, { depth: null }));
  // client.on(PACKETS.carTelemetry, console.log);
  // client.on(PACKETS.carTelemetry, console.log);
  // client.on(PACKETS.lobbyInfo, console.log);

  client.on(PACKETS.event, (packet) => {
    raceData = trackEvents(packet, raceData);
    persistData(raceData);
  });
  client.on(PACKETS.lapData, (packet) => {
    raceData = trackLapData(packet, raceData, carStatus);
    persistData(raceData);
  });

  client.on(PACKETS.participants, (packet) => {
    raceData = trackParticipants(packet, raceData);
    persistData(raceData);
  });

  client.on(PACKETS.carStatus, (packet) => {
    carStatus = packet;
  });
  client.on(PACKETS.finalClassification, (packet) => {
    raceData = trackFinalClassification(packet, raceData);
    persistData(raceData, true);
    console.info(`Session ${raceData.sessionId} complete`);
    resetData();
  });

  // to start listening:
  client.start();

  console.info(`Streaming data to ${path}`);

  // TODO: On start, load file if it exists for this session?

  // and when you want to stop:
  // client.stop();
};
